### ISSUE TITLE
Tiller pods can't connect to k8s apiserver

### ISSUE THREAD

**Issue by willise** [kubernetes/helm#2464]
logs:
```
(kube-master) tiller # kcm logs tiller-deploy-3354596499-4q16f
Cannot initialize Kubernetes connection: Get http://localhost:8080/api: dial tcp [::1]:8080: getsockopt: connection refused
```
This issue has been raised in #1591 #1791.

In #1791, there gets a resolution(thank @iamzhout) of adding **KUBERNETES_MASTER** in pods env  but nobody explains the root cause of the issue.

I installed the k8s cluster manually and it seems that  tiller pod can't get **KUBERNETES_MASTER** from env as this item doesn't exist in the **env** at all. 

No docs are found about the relation between k8s cluster and the configuration.  I don't  think it is a good idea to add the env item manually(and how to `helm init` with the configuration???)

Is there anybody who can offer the help? 

Thanks anyway (^.^).

**Comment by technosophos** [comment]
@willise You might want to ask about this in the Slack channel if possible. It is very likely a configuration issue with Kubernetes, and people there may be able to help quickly.

**Comment by bacongobbler** [comment]
Agreed. This looks to be more of a misconfigured kubernetes cluster causing the KUBERNETES_SERVICE_HOST environment variable to either be pointing in the wrong location or it doesn't exist at all. Tiller needs that envvar in the pod to connect to the kubernetes apiserver.

I don't think there's anything actionable here in helm itself, save perhaps some docs clarification on how tiller communicates with the apiserver.

**Comment by willise** [comment]
@technosophos Ooops... Kind of your reminding and I'll note that.

**Comment by willise** [comment]
@bacongobbler **KUBERNETES_SERVICE_HOST** exists in **env** indeed but it's value is a cluster IP like **10.254.0.1**. If I don't set **KUBERNETES_MASTER**, tiller is unable to connect apiserver port 8080. Only if it is set to master physical address like **192.168.56.101**, it's ok. 

So the problem may be  tiller can't get the correct apiserver address from k8s when tiller initializes the k8s apiserver connection.

**Comment by allamand** [comment]
I've got a similar error where tiller retrieve the correct ip and port but can't connect due to tls verification (i use skip-tls-verify on my kubernetes.

```
2017-05-30T10:52:56.35985785Z Cannot initialize Kubernetes connection: Get https://10.100.200.1:443/api: x509: cannot validate certificate for 10.100.200.1 because it doesn't contain any IP SANs
```
is there a way to ask tiller to skip this verification ?

**Comment by penfree** [comment]
I meet the same problem, it was due to the version compatibility of k8s and helm.   my k8s version is 1.3, when i use helm 2.0.2, everything is ok. i'm not sure what's the highest compatible version for k8s 1.3

**Comment by mvernimmen** [comment]
@bacongobbler I'm having the same issue here, while KUBERNETES_SERVICE_HOST is set correctly:
```
$ kubectl exec -it tiller-deploy-1490305151-6jq17 -n integrations-test -- /bin/sh
/ # env | grep -i kube
KUBERNETES_PORT=tcp://10.96.0.1:443
KUBERNETES_SERVICE_PORT=443
KUBERNETES_PORT_443_TCP_ADDR=10.96.0.1
KUBERNETES_PORT_443_TCP_PORT=443
KUBERNETES_PORT_443_TCP_PROTO=tcp
KUBERNETES_SERVICE_PORT_HTTPS=443
KUBERNETES_PORT_443_TCP=tcp://10.96.0.1:443
KUBERNETES_SERVICE_HOST=10.96.0.1
```

To fix the problem I set the KUBERNETES_MASTER in the running tiller pod and tried again, but when doing an install, the problem remains:
```
$ kubectl logs tiller-deploy-1490305151-6jq17 -n integrations-test
...
[tiller] 2017/12/28 13:34:09 preparing install for rmlater
[storage] 2017/12/28 13:34:09 getting release history for "rmlater"
[storage/driver] 2017/12/28 13:34:09 query: failed to query with labels: Get http://localhost:8080/api/v1/namespaces/integrations-test/configmaps?labelSelector=NAME%3Drmlater%2COWNER%3DTILLER: dial tcp 127.0.0.1:8080: getsockopt: connection refused
[tiller] 2017/12/28 13:34:09 failed install prepare step: Get http://localhost:8080/version: dial tcp 127.0.0.1:8080: getsockopt: connection refused
```

And trying the install with `--kube-context integrations@kube.t.nl02` or with KUBECONFIG set to the path of my .kube/config file did not resolve the problem either.

According to the kubernetes manual, the best practice is to resolve the kubernetes (fqdn) hostname (https://kubernetes.io/docs/tasks/access-application-cluster/access-cluster/#accessing-the-api-from-a-pod), is that not what is being done here?
Why is the KUBERNETES_SERVICE_HOST not being used to access the API?
This is all with version 2.7.2

I'm probably missing something obvious, as are some of the other people having this problem. Could you, or someone else please have a look into this?

**Comment by mvernimmen** [comment]
I have found 1 possible cause for this problem.
On a cluster set up with kubespray, running version 1.8.3, helm init and helm install work when executed by a user when deploying to a specific namespace. But only when the serviceaccount name is 'tiller'. When using serviceaccount name 'tiller-integrations-acc', the 127.0.0.1 error appears.
I've pasted my steps here: https://pastebin.com/TwCypx4Q
Did I do it wrong or did I run into a bug?

However, this did not solve the problem on a cluster set up using kubeadm and running k8s version 1.7.8, there the 127.0.0.1 API problem remains despite using the serviceaccount name 'tiller'. So perhaps there are multiple possible causes for the error message.

**Comment by mattus** [comment]
Had something very similar to this, our eventual fix was to ensure _automountServiceAccountToken_ on the _ServiceAccount_ running tiller was set to true. We were creating the service account with terraform which was defaulting to false for reasons discussed in https://github.com/terraform-providers/terraform-provider-kubernetes/issues/38. 

Manually creating the service account with kubectl does not set this value (it defaults to true when unset).

**Comment by mvernimmen** [comment]
Hi Mattus, thank you for suggesting that. I tried it, but unfortunately for me it doesn't fix the problem. installing charts when using a service account name other than 'tiller' results in helm not being able to find 'tiller'.

**Comment by bacongobbler** [comment]
According to #3816 and a comment in #1791, it appears that the missing environment variable that is injected from cloud providers and [assumed to be present as per client-go is KUBERNETES_MASTER](https://github.com/kubernetes/client-go/blob/d74b6802c2c97ec8bb9c6c169de982e13e84d01a/tools/clientcmd/client_config.go#L47-L54), not KUBERNETES_SERVICE_HOST. I'm not sure what the resolution is, but this assumption is embedded in client-go and is consistent across all the major cloud providers, so it's probably something missing in the bare metal cluster setup guide which breaks this assumption. :/

tl;dr this is a bare metal cluster doc issue, not necessarily something wrong with helm's assumptions about the running environment.

**Comment by fossxplorer** [comment]
@mattus Thanks a lot, i was stuck for ~ 3 days with this at work trying to deploy a k8s cluster. This should really be documented somewhere. 
What i did to solve the issue was:
- kubectl --namespace=kube-system edit deployment/tiller-deploy and changed automountServiceAccountToken to true. 
Then 'helm list' was giving me:
Error: configmaps is forbidden: User "system:serviceaccount:kube-system:default" cannot list configmaps in the namespace "kube-system"
That was fixed with solution from https://github.com/kubernetes/helm/issues/2687:
-  kubectl --namespace=kube-system create clusterrolebinding add-on-cluster-admin --clusterrole=cluster-admin --serviceaccount=kube-system:default

**Comment by bacongobbler** [comment]
@fossexplorer if you want to document your findings we'd really appreciate it!

**Comment by johnhamelink** [comment]
FYI this is happening when I use terraform to build an AKS k8s cluster in Azure. This issue gave a clue: https://github.com/terraform-providers/terraform-provider-kubernetes/issues/38

Editing the deploy like @fossxplorer mentions seems to help me. Here's a handy one-liner to do just that:

```
kubectl -n kube-system patch deployment tiller-deploy -p '{"spec": {"template": {"spec": {"automountServiceAccountToken": true}}}}'
```

**Comment by kfox1111** [comment]
I just hit this with the newest version on the newest minikube. :/

**Comment by kfox1111** [comment]
editing tiller-deployment,
spec.template.spec.automountServiceAccountToken and setting it to true (it was false) seemed to help.

**Comment by milescrabill** [comment]
I ran into this on a fresh GKE cluster that I manually enabled RBAC on per Google's docs [here](https://cloud.google.com/kubernetes-engine/docs/how-to/role-based-access-control).

I had already installed Helm, so things broke down when I enabled RBAC. After creating a Serviceaccount and ClusterRoleBinding for tiller, upgrading Helm with `helm init --service-account tiller --upgrade` was still producing the error above: `Error: Get http://localhost:8080/api/v1/namespaces/kube-system/configmaps?labelSelector=OWNER%!D(MISSING)TILLER: dial tcp [::1]:8080: connect: connection refused`.

As others have said, updating the tiller deployment to change `automountServiceAccountToken: true` fixed my issue.

**Comment by andrewgdavis** [comment]
Alternatively, one is able to install with the default service account specified:
`helm init --service-account default`
which will install tiller to the kube-system namespace.

**Comment by kfox1111** [comment]
looks like minikube in the newest version deploys with rbac enabled. That is another difference.

**Comment by bacongobbler** [comment]
If you would be so kind as to test #3990, that would be appreciated. Seems like there was a regression in 2.9.

**Comment by bacongobbler** [comment]
alternative PR: #3991

**Comment by miroadamy** [comment]
The solution presented above by fossxplorer works for version v0.25.2 of minikube, but same sequence seems to be failing for v0.26.1.

**Comment by AmazingTurtle** [comment]
@miroadamy minikube v0.26.* seems to be broken at all, certificates fucked up. v0.25.2 is safe to use right now

**Comment by bacongobbler** [comment]
This has been fixed in helm v2.9.1, which is available here: https://github.com/kubernetes/helm/releases/v2.9.1

Thanks everyone!

**Comment by AmazingTurtle** [comment]
Upgraded vom v2.7.2 to v2.9.1 using `helm init --upgrade` (also tried `helm init --service-account default --upgrade`). Still had to apply fix https://github.com/kubernetes/helm/issues/2464#issuecomment-381101015 (automountServiceAccountToken helped).

Worth to mention: Running minikube with rbac enabled (https://gist.github.com/F21/08bfc2e3592bed1e931ec40b8d2ab6f5)

**Comment by ramyala** [comment]
@bacongobbler  Can we reopen this issue. If your serviceaccount has automountServiceAccountToken=false, v2.9.1 will fail with continue to fail. The pod configuration for tiller-deploy should allow serviceaccount mounts by explicitly specifying automountServiceAccountToken:true.

**Comment by ramyala** [comment]
I sent out a PR which fixes this issue and am able to get terraform service_accounts to work with helm.

PTAL @ #4229

**Comment by ramyala** [comment]
did you try #4229? It resolved the issues in my case

**Comment by paultiplady** [comment]
I hit this issue on a fresh install of Kubernetes under Docker for Mac.

The above suggestion to set automountServiceAccountToken to true resolved the issue for me.

**Comment by johnmshields** [comment]
This might help others that come upon this issue when working with terraform. The following terraform configuration was able to both create the service account and properly initialize tiller. The key is the `override` parameter to the helm provider.

```
data "google_client_config" "current" {}

provider "kubernetes" {
  load_config_file = false

  host                   = "https://${var.master_ip}"
  token                  = "${data.google_client_config.current.access_token}"
  cluster_ca_certificate = "${base64decode(var.cluster_ca_certificate)}"
}

provider "helm" {
  namespace       = "kube-system"
  service_account = "${kubernetes_service_account.tiller.metadata.0.name}"
  override        = ["spec.template.spec.automountserviceaccounttoken=true"]

  kubernetes {
    host                   = "https://${var.master_ip}"
    token                  = "${data.google_client_config.current.access_token}"
    cluster_ca_certificate = "${base64decode(var.cluster_ca_certificate)}"
  }
}

resource "kubernetes_service_account" "tiller" {
  metadata {
    name      = "tiller"
    namespace = "kube-system"
  }
}

resource "kubernetes_cluster_role_binding" "tiller" {
  metadata {
    name = "tiller"
  }

  # TODO: give specific permissions
  role_ref {
    name = "cluster-admin"
    kind = "ClusterRole"
  }

  subject {
    kind      = "ServiceAccount"
    name      = "tiller"
    namespace = "kube-system"
  }
}
```

**Comment by ttaranov** [comment]
seems to be the same issue - docker for mac, with kubernetes enabled locally, Helm fails to list or otherwise work beyond init with the i/o timeout:

`
$ helm list
Error: Get https://10.96.0.1:443/api/v1/namespaces/kube-system/configmaps?labelSelector=OWNER%!D(MISSING)TILLER: dial tcp 10.96.0.1:443: i/o timeout

$ helm version
Client: &version.Version{SemVer:"v2.13.1", GitCommit:"618447cbf203d147601b4b9bd7f8c37a5d39fbb4", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.13.1", GitCommit:"618447cbf203d147601b4b9bd7f8c37a5d39fbb4", GitTreeState:"clean"}
`
tried installing helm with specific service account and net `helm init --service-account tiller --net-host` - to no avail

**Comment by grajeshwaran** [comment]
Just upgrade helm to version 3.
